pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -0,0 +1,230 @@
+// timeout in check without adding to it on every op. If we didn't initiate
+// a wait or if timeout was 0 then wait channel will be nil.
+// Returned allocation has its use counter set to 1.
+func (f *FeedBudget) Get(
+	ctx context.Context, amount int64, timeout time.Duration,
+) (*SharedBudgetAllocation, <-chan time.Time, error) {
+	alloc, err := f.tryAlloc(ctx, amount)
+	if err == nil {
+		return getPooledBudgetAllocation(alloc), nil, nil
+	}
+	// If we failed to allocate immediately, then we need to wait on replenish
+	// channel to track Return calls from consumed events.
+	var timeC <-chan time.Time
+	if timeout > 0 {
+		timeC = time.After(timeout)
+	}
+	for {
+		select {
+		case <-f.replenishC:
+			alloc, err = f.tryAlloc(ctx, amount)
+			if err == nil {
+				return getPooledBudgetAllocation(alloc), timeC, nil
+			}
+		case <-timeC:
+			// TODO(oleg): should we try to allocate again in case we are faced
+			// exhausted memory pool and someone returned memory to the pool?
+			f.metrics.RangeFeedBudgetExhausted.Inc(1)
+			return nil, timeC, errors.New("failed to allocate mem budget")
+		case <-ctx.Done():
+			return nil, timeC, nil
+		case <-f.stopC:
+			// We are already stopped, current allocation is already freed so, do
+			// nothing.
+			return nil, timeC, nil
+		}
+	}
+}
+
+// Function to allocate from budget or if there's nothing there borrow from
+// emergency ether.
+func (f *FeedBudget) tryAlloc(ctx context.Context, amount int64) (SharedBudgetAllocation, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.mu.closed {
+		return SharedBudgetAllocation{}, errors.New("budget is already released")
+	}
+	if f.mu.oversize {
+		return SharedBudgetAllocation{}, errors.New("failed to allocate mem budget")
+	}
+	err := f.mu.memBudget.Grow(ctx, amount)
+	if err == nil {
+		return SharedBudgetAllocation{size: amount, refCount: 1, feed: f}, nil
+	}
+	// If we didn't use any budget yet and not currently sending an oversized
+	// event then we can send one.
+	if f.mu.memBudget.Used() == 0 {
+		f.metrics.RangeFeedOverBudgetEvents.Inc(1)
+		f.metrics.RangeFeedOverBudgetAllocation.Inc(amount)
+		f.mu.oversize = true
+		// We are in override mode, we return 0 size to avoid returning it to
+		// budget later as we didn't borrow anything.
+		return SharedBudgetAllocation{size: 0, refCount: 1, feed: f}, nil
+	}
+	return SharedBudgetAllocation{}, err
+}
+
+// Return returns amount to budget.
+func (f *FeedBudget) returnAllocation(ctx context.Context, amount int64) {
+	f.mu.Lock()
+	if f.mu.closed {
+		f.mu.Unlock()
+		return
+	}
+	if f.mu.oversize {
+		f.metrics.RangeFeedOverBudgetAllocation.Inc(-amount)
+	}
+	f.mu.oversize = false
+	if amount > 0 {
+		f.mu.memBudget.Shrink(ctx, amount)
+	}
+	f.mu.Unlock()
+	select {
+	case f.replenishC <- struct{}{}:
+	default:
+	}
+}
+
+// Close frees up all allocated budget and prevents any further allocations.
+// Safe to call on nil budget.
```

Comment:
I think you'll want to put this inside the `closed` check too, since closing a channel twice will panic.

Do you know about `sync.Once`? It's usually used for these sorts of close patterns, e.g. add a `closeOnce sync.Once` to the struct and then do:

```
f.closeOnce.Do(func() {
  f.mu.Lock()
  f.mu.memBudget.Close(ctx)
  f.mu.Unlock()
  close(f.stopC)
})
```

Reply:
That's neat! We still need a closed flag though as we play safe and drop all allocation/return attempts if budget was closed. Just in case shutdown is out of order.